### PR TITLE
Added `StoreProduct.getEligibleDiscounts()` and `Purchases.getEligibleDiscounts(forProduct:)`

### PR DESCRIPTION
**Pull request by NachoSoto**
This simplifies the API and allows users to use only one method to determine which `StoreProductDiscount`s to display in the UI.

### CODE REVIEW

File: Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift
Comment (NachoSoto) [inline Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift:234]: Before anyone asks: I don't think there's any way to simplify this making it more functional because it's using `AsyncStream`.

Comment (joshdholtz) [reply Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift:234]: Yeah, I _think_ this is as clean as it gets until the concurrency gets a more functional API 🤷‍♂️

File: Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift
Comment (joshdholtz) [inline Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift:200]: Is passing in a `Purchases` reference a new public API concept? I

It seems like might expect an API that is something like 👇 

```swift
Purchases.getEligibleDiscounts(forStoreProduct: storeProduct)
```

Thoughts? 😊

Comment (NachoSoto) [reply Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift:200]: I agree it's slightly weird, but it improves discoverability. When users type `product.discounts` they'll see the autocomplete for both.
What do you think?

File: Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift
Comment (aboedo) [inline Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift:157]: missing one backtick at the end of this line

File: Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift
Comment (aboedo) [inline Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift:195]: kinda feel like `Asynchronously` here is redundant - the API already says `async`, and most of our methods are async anway

File: Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift
Comment (aboedo) [reply Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift:200]: 🤔 +1 to Josh's point, I would have expected the API to look exactly like what he described. 
That said, I do see the advantage of having access to it from `discounts`. 

Could we have both? That way, if you're used to using `Purchases` as the entrypoint, it won't feel weird, but you can also use the convenience of getting the eligible discounts easily. 

Alternatively, we could go ahead and use the `shared` instance of `Purchases` ourselves, since that's the one we're supposed to be using in the first place. 

Note that in either case we do have to think about what happens if this is somehow called with a non-configured instance of `Purchases`

File: Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift
Comment (NachoSoto) [reply Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift:157]: Oops thanks

File: Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift
Comment (NachoSoto) [reply Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift:195]: That's fair. I wanted to make it clear that this made API requests, but I guess you're right.

File: Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift
Comment (beylmk) [inline Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift:213]: since we end up checking if nil later anyway, maybe it'd be clearer to just return `eligibility.isEligible` and check that?

File: Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift
Comment (NachoSoto) [reply Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift:200]: > Alternatively, we could go ahead and use the shared instance of Purchases ourselves, since that's the one we're supposed to be using in the first place.
> Note that in either case we do have to think about what happens if this is somehow called with a non-configured instance of Purchases

Oooh of course! I don't like implicitly depending on a global, but in the case of this SDK it's basically unusable without calling `configure`.
I'll also add a note to indicate that it's required 👍

File: Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift
Comment (beylmk) [reply Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift:195]: was thinking maybe we should rename this to `getEligibleDiscounts` to make more clear that it is async, but then i realized we don't really follow that convention for async functions, only for those with completion blocks? is that by design?

Comment (NachoSoto) [reply Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift:195]: Actually I think that's a great way to convey that this is a potentially "expensive" operation instead of a simpler property getter 👍

**Comment by beylmk** [comment]
❤️  this API... i wonder if we can bubble it up toe the `Package` level too?

**Comment by NachoSoto** [comment]
Yeah, maybe we could add a protocol `PackageOrProduct` (obviously with a better name) so we can share some of the same interface between the 2? Then easily forward the implementation from Package to Product.

File: Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift
Comment (NachoSoto) [reply Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift]: Changed comment and method name.

**Comment by NachoSoto** [comment]
Filed https://app.shortcut.com/revenuecat/story/12875/consider-exposing-more-product-methods-to-package

File: Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift
Comment (NachoSoto) [reply Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift]: Done.

Comment (aboedo) [reply Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift]: Awesome! ❤️ 

@NachoSoto any thoughts on this bit? 

> Could we have both? That way, if you're used to using Purchases as the entrypoint, it won't feel weird, but you can also use the convenience of getting the eligible discounts easily.

Comment (NachoSoto) [reply Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift]: Oh yes of course! Done!

**Comment by NachoSoto** [comment]
This should be ready. I also added the methods to `APITester`.

File: Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift
Comment (aboedo) [inline Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift:161]: we have a new format for related symbols now with #1253, but we can update the docs here after the PR is merged so we don't hold this one any longer

File: Purchases/Purchasing/StoreKitAbstractions/PromotionalOffer.swift
Comment (aboedo) [inline Purchases/Purchasing/StoreKitAbstractions/PromotionalOffer.swift:62]: ❤️

File: Purchases/Purchasing/Purchases.swift
Comment (aboedo) [inline Purchases/Purchasing/Purchases.swift:1237]: 👏 
We should also document the `- parameter` but we can do that as a follow-up

**Review by aboedo — APPROVED** [review]
This looks awesome! I'm super happy about this API

File: Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift
Comment (NachoSoto) [reply Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift]: Changed! Could you take another look to make sure I didn't mess up the new format?

File: Purchases/Purchasing/Purchases.swift
Comment (NachoSoto) [reply Purchases/Purchasing/Purchases.swift]: Added!

File: Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift
Comment (beylmk) [inline Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift:159]:
```suggestion
    /// - Note: the current user may or may not be eligible for some of these.
```

File: Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift
Comment (beylmk) [inline Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift:198]: sorry if i missed a discussion on this...but should we error if purchases isn't initialized?

Comment (NachoSoto) [reply Purchases/Purchasing/StoreKitAbstractions/StoreProduct.swift:198]: I could go either way. I think it's probably better to crash, just like calling `Purchases.shared` would?

**Review by joshdholtz — APPROVED** [review]
Thank you for adding the adding onto `Purchases` 🙌

**Pull request merged**
